velox/functions/prestosql/GeometryUtils.h

```diff
@@ -0,0 +1,685 @@
+          coordinates,
+          partIndexes.back(),
+          coordinates->size(),
+          shellPart.back());
+    }
+  }
+
+  static void canonicalizePolygonCoordinates(
+      const std::unique_ptr<geos::geom::CoordinateSequence>& coordinates,
+      int start,
+      int end,
+      bool isShell) {
+    bool isClockwiseFlag = isClockwise(coordinates, start, end);
+
+    if ((isShell && !isClockwiseFlag) || (!isShell && isClockwiseFlag)) {
+      reverse(coordinates, start, end);
+    }
+  }
+
+  template <typename T>
+  static void writeGeometryCollection(
+      const geos::geom::Geometry* collection,
+      SliceOutput<T>& output) {
+    output.writeByte(
+        static_cast<uint8_t>(GeometrySerializationType::GEOMETRY_COLLECTION));
+
+    for (size_t geometryIndex = 0;
+         geometryIndex < collection->getNumGeometries();
+         ++geometryIndex) {
+      auto* geometry = collection->getGeometryN(geometryIndex);
+      // Use a temporary buffer to serialize the geometry and calculate its
+      // length
+      std::string tempBuffer;
+      SliceOutput tempOutput(tempBuffer);
+
+      serialize(geometry, tempOutput);
+
+      int32_t length = static_cast<int32_t>(tempBuffer.size());
+      output.writeInt(length);
+      output.writeBytes(tempBuffer.data(), tempBuffer.size());
+    }
+  }
+
+  inline static thread_local auto GEOMETRY_FACTORY =
+      geos::geom::GeometryFactory::create();
+
+  static void skipEsriType(SliceInput& input) {
+    input.skip(4); // Esri type is an integer
+  }
+
+  static void skipEnvelope(SliceInput& input) {
+    input.skip(4 * 8); // Envelopes are 4 doubles (minX, minY, maxX, maxY)
+  }
+
+  static geos::geom::Coordinate readCoordinate(SliceInput& input) {
+    double x = input.readDouble();
+    double y = input.readDouble();
+    return {x, y};
+  }
+
+  static std::unique_ptr<geos::geom::CoordinateSequence> readCoordinates(
+      SliceInput& input,
+      int count) {
+    auto coords = std::make_unique<geos::geom::CoordinateSequence>(count, 2);
+    for (int i = 0; i < count; ++i) {
+      coords->setAt(readCoordinate(input), i);
+    }
+    return coords;
+  }
+
+  static std::unique_ptr<geos::geom::Point> readPoint(SliceInput& input) {
+    geos::geom::Coordinate coordinate = readCoordinate(input);
+    if (std::isnan(coordinate.x) || std::isnan(coordinate.y)) {
+      return GEOMETRY_FACTORY->createPoint();
+    }
+    return std::unique_ptr<geos::geom::Point>(
+        GEOMETRY_FACTORY->createPoint(coordinate));
+  }
+
+  static std::unique_ptr<geos::geom::Geometry> readMultiPoint(
+      SliceInput& input) {
+    skipEsriType(input);
+    skipEnvelope(input);
+    int pointCount = input.readInt();
+    auto coords = readCoordinates(input, pointCount);
+    std::vector<std::unique_ptr<geos::geom::Point>> points;
+    for (size_t i = 0; i < coords->size(); ++i) {
+      points.push_back(
+          std::unique_ptr<geos::geom::Point>(GEOMETRY_FACTORY->createPoint(
```

Comment:
is the memory returned by GEOMETRY_FACTORY->createPoint() owned by it? or is the caller responsible for it?